### PR TITLE
task: schedule sticky tasks correctly with _wait2

### PR DESCRIPTION
**Pull request by vtjnash**
The `_wait2` function is similar to calling `schedule + wait` from
another `@async` task, but optimized, so we want to observe the same
side-effects of making the task sticky to the correct thread (and not
accidentally making it sticky to the later task that handles the event).

Refs #41334 (75858d73322)

### CODE REVIEW

**Review by tkf — COMMENTED** [review]
Uh... this is unfortunate but makes sense to me. How about adding something like

```julia
"""
    _coschedule_parent_if_required(t::Task) -> Bool

Make `current_task()` sticky if `t` is a newly scheduled sticky task and return `true`.

t.sticky && threadid(t) == 0 is a task that needs to be co-scheduled with
the parent task. If the parent (current_task) is not sticky we must
set it to be sticky.
XXX: Ideally we would be able to unset this
"""
function _coschedule_parent_if_required(t::Task)
    (t.sticky && Threads.threadid(t) == 0) || return false
    current_task().sticky = true
    tid = Threads.threadid()
    ccall(:jl_set_task_tid, Cvoid, (Any, Cint), t, tid-1)
    return true
end
```

to avoid repeating the tricky code?

**Comment by vtjnash** [comment]
The code shouldn't need to be duplicated more, and apparently doesn't have a good name for splitting out in that helper

**Review by vchuravy — APPROVED** [review]
For the record. I hate this :)

File: base/task.jl [hunk 1]
```diff
@@ -306,6 +306,18 @@ function _wait2(t::Task, waiter::Task)
...
+            # bit, even if we aren't calling `schedule` due to this early-return
+            if waiter.sticky && Threads.threadid(waiter) == 0
+                # Issue #41324
+                # t.sticky && tid == 0 is a task that needs to be co-scheduled with
+                # the parent task. If the parent (current_task) is not sticky we must
+                # set it to be sticky.
+                # XXX: Ideally we would be able to unset this
+                current_task().sticky = true
+                tid = Threads.threadid()
+                ccall(:jl_set_task_tid, Cvoid, (Any, Cint), waiter, tid-1)
```
Comment (kpamnany) [inline base/task.jl:319]: @vtjnash and @vchuravy: I think this is wrong. Here, `waiter` is waiting on `t`, not on `current_task()` which can be completely different (e.g. see [bind](https://github.com/JuliaLang/julia/blob/master/base/channels.jl#L256)).

Apparently, we want to pin `t` and `waiter` to the same thread. But I'm not sure how this follows from the original problem: #41324. It's perfectly reasonable for a task in the interactive threadpool to wait on a task in the default threadpool, but this makes that unachievable.

I know it's been a while, but can either of you remember why this was added after #41334?

Comment (vtjnash) [reply base/task.jl:319]: `_wait2` is defined as semantically a call to `schedule` the Task to immediately call `wait` on some condition variable. Hence why it has the same side-effects as calling `schedule`: it should pin `current_task` and `waiter` to one thread, but not `t` (which will have completed before `waiter` runs anyways).

Comment (kpamnany) [reply base/task.jl:319]: That task `t` will have completed before `waiter` runs makes sense. But _this_ `_wait2` as opposed to the one in `condition.jl` has nothing to do with `Condition`s. `t` is a `Task` here, not a `Condition`. The semantics of this `_wait2` seem unrelated to `current_task()`.

**Pull request merged**